File: server/temporal.py
"""Temporal graph contract shared with web/js/kg.js through offline fixtures."""
from collections import defaultdict


GENERIC = set('父亲 母亲 爸爸 妈妈 爹 娘 儿子 女儿 丈夫 妻子 太太 夫人 先生 老爷 小姐 姑娘 少女 女孩 少年 男孩 青年 少爷 医生 大夫 校长 老师 神父 堂长 老板 老板娘 仆人 女仆 老头子 老太太 老头 孩子 哥哥 姐姐 弟弟 妹妹 叔叔 伯父 舅舅 姑妈 姨妈 祖父 祖母 爷爷 奶奶 外公 外婆 公公 婆婆 岳父 岳母 丈人 主人 客人 新娘 新郎 新娘子 新夫人 寡妇 邻居 朋友 同学 学生 病人 大人 老人 年轻人 女人 男人 闺女 媳妇 老婆 老公 东家 女婿 未婚女婿 未婚妻 未婚夫 儿媳'.split())


def fold(log, pos):
    import re
    people, merges, relations, events, recaps = {}, {}, [], [], []
    early = defaultdict(list)
    saga = ''

    def apply(r):
        kind = r['t']
        pid = r.get('id')
        if kind in ('name', 'alias', 'profile', 'attr', 'imp') and pid not in people:
            early[pid].append(r)
            return
        if kind == 'person':
            people[pid] = {'id': pid, 'name': r['name'], 'aliases': [], 'tagline': r.get('intro', ''),
                           'bio': '', 'attrs': {}, 'attr_history': defaultdict(list),
                           'imp': r.get('imp', 1), 'n': 0, 'events': []}
            for held in early.pop(pid, []):
                apply(held)
        elif kind == 'name':
            if people[pid]['name'] != r['name']:
                people[pid]['aliases'].append(people[pid]['name'])
                people[pid]['name'] = r['name']
        elif kind == 'alias':
            people[pid]['aliases'].append(r['alias'])
        elif kind == 'merge':
            merges[r['from']] = r['into']
        elif kind == 'profile':
            people[pid]['tagline'] = r.get('tagline') or people[pid]['tagline']
            people[pid]['bio'] = r.get('bio') or people[pid]['bio']
        elif kind == 'attr':
            people[pid]['attr_history'][r['key']].append({'v': r['value'], 'p': r['p']})
        elif kind == 'rel':
            relations.append(r)
        elif kind == 'event':
            events.append(r)
        elif kind == 'imp':
            people[pid]['imp'] = r['imp']
        elif kind == 'cnt':
            for who, count in r['c'].items():
                if who in people:
                    people[who]['n'] += count
        elif kind == 'recap':
            recaps.append(r)

    for row in log:
        if row['p'] > pos:
            break
        if row['t'] == 'saga':
            saga = row['text']
        else:
            apply(row)

    def canon(pid):
        seen = set()
        while pid in merges and pid not in seen:
            seen.add(pid)
            pid = merges[pid]
        return pid

    for src_id in merges:
        target_id = canon(src_id)
        if src_id == target_id or src_id not in people or target_id not in people:
            continue
        src, target = people.pop(src_id), people[target_id]
        if target['name'] in GENERIC and src['name'] not in GENERIC:
            target['name'] = src['name']
        target['aliases'].extend([src['name']] + src['aliases'])
        target['n'] += src['n']
        if not target['bio']:
            target['bio'] = src['bio']
        for key, values in src['attr_history'].items():
            target['attr_history'][key] = values + target['attr_history'][key]
    names = {p['name']: p['id'] for p in people.values()}
    for person in people.values():
        person['aliases'] = list(dict.fromkeys(x for x in person['aliases'] if x and x != person['name']
            and x not in GENERIC and re.sub(r'^[他她的老小未]+', '', x) not in GENERIC
            and not re.search(r'的|夫妇|夫妻|们|一家|俩', x)
            and not (x in names and names[x] != person['id'])))
        for key, values in person['attr_history'].items():
            values.sort(key=lambda x: x['p'])
            person['attrs'][key] = values[-1]['v']
    for event in events:
        for pid in {canon(x) for x in event['who']}:
            if pid in people:
                people[pid]['events'].append(event)
    rels = {}
    for record in relations:
        a, b = canon(record['a']), canon(record['b'])
        if a == b or a not in people or b not in people:
            continue
        key = '|'.join(sorted((a, b))) + ('|' + record['family'] if record.get('family') else '')
        previous = rels.get(key)
        row = dict(record, a=a, b=b)
        if a > b:
            row.update(a=b, b=a, a_is=record.get('b_is', ''), b_is=record.get('a_is', ''))
        if previous:
            reverse = previous['a'] != row['a']
            for field, oldfield in [('a_is', 'b_is' if reverse else 'a_is'),
                                    ('b_is', 'a_is' if reverse else 'b_is'), ('desc', 'desc'), ('status', 'status')]:
                row[field] = row.get(field) or previous.get(oldfield, '')
            row['history'] = previous.get('history', []) + [{k: v for k, v in previous.items() if k != 'history'}]
        else:
            row['history'] = []
        rels[key] = row
    return {'people': people, 'rels': rels, 'events': events, 'saga': saga, 'recaps': recaps}

File: core/lib/src/server/temporal.dart
/// Server evidence graph (`server/temporal.py`), distinct from reader UI data.
library;

import '../fold/fold.dart' show Json, generic;

export '../fold/fold.dart' show Json;

Json foldEvidence(List<Json> log, int pos) {
  final Map<String, Json> people = <String, Json>{};
  final Map<String, String> merges = <String, String>{};
  final List<Json> relations = <Json>[];
  final List<Json> events = <Json>[];
  final List<Json> recaps = <Json>[];
  final Map<String, List<Json>> early = <String, List<Json>>{};
  String saga = '';
  void apply(Json r) {
    final String kind = r['t']! as String;
    final String? id = r['id'] as String?;
    if (<String>['name', 'alias', 'profile', 'attr', 'imp'].contains(kind) &&
        !people.containsKey(id)) {
      early.putIfAbsent(id!, () => <Json>[]).add(r);
      return;
    }
    switch (kind) {
      case 'person':
        people[id!] = <String, Object?>{
          'id': id,
          'name': r['name'],
          'aliases': <String>[],
          'tagline': r['intro'] ?? '',
          'bio': '',
          'attrs': <String, Object?>{},
          'attr_history': <String, Object?>{},
          'imp': r['imp'] ?? 1,
          'n': 0,
          'events': <Json>[],
        };
        for (final Json held in early.remove(id) ?? <Json>[]) {
          apply(held);
        }
      case 'name':
        if (people[id]!['name'] != r['name']) {
          (people[id]!['aliases']! as List<String>).add(
            people[id]!['name']! as String,
          );
          people[id]!['name'] = r['name'];
        }
      case 'alias':
        (people[id]!['aliases']! as List<String>).add(r['alias']! as String);
      case 'merge':
        merges[r['from']! as String] = r['into']! as String;
      case 'profile':
        for (final String key in <String>['tagline', 'bio']) {
          if (r[key] != null && r[key] != '') people[id]![key] = r[key];
        }
      case 'attr':
        final Json history = people[id]!['attr_history']! as Json;
        (history.putIfAbsent(r['key']! as String, () => <Json>[])!
                as List<Json>)
            .add(<String, Object?>{'v': r['value'], 'p': r['p']});
      case 'rel':
        relations.add(r);
      case 'event':
        events.add(r);
      case 'imp':
        people[id]!['imp'] = r['imp'];
      case 'cnt':
        for (final MapEntry<String, Object?> e in (r['c']! as Json).entries) {
          if (people.containsKey(e.key)) {
            people[e.key]!['n'] =
                (people[e.key]!['n']! as num) + (e.value! as num);
          }
        }
      case 'recap':
        recaps.add(r);
    }
  }

  for (final Json row in log) {
    if ((row['p']! as num) > pos) break;
    if (row['t'] == 'saga') {
      saga = row['text']! as String;
    } else {
      apply(row);
    }
  }
  String canon(String id) {
    final Set<String> seen = <String>{};
    while (merges.containsKey(id) && seen.add(id)) {
      id = merges[id]!;
    }
    return id;
  }

  for (final String id in merges.keys) {
    final String targetId = canon(id);
    if (id == targetId ||
        !people.containsKey(id) ||
        !people.containsKey(targetId))
      continue;
    final Json source = people.remove(id)!;
    final Json target = people[targetId]!;
    if (generic.contains(target['name']) && !generic.contains(source['name']))
      target['name'] = source['name'];
    (target['aliases']! as List<String>).addAll(<String>[
      source['name']! as String,
      ...source['aliases']! as List<String>,
    ]);
    target['n'] = (target['n']! as num) + (source['n']! as num);
    if (target['bio'] == '') target['bio'] = source['bio'];
    final Json history = target['attr_history']! as Json;
    for (final MapEntry<String, Object?> e
        in (source['attr_history']! as Json).entries) {
      history[e.key] = <Json>[
        ...e.value! as List<Json>,
        ...(history[e.key] as List<Json>?) ?? <Json>[],
      ];
    }
  }
  final Map<Object?, Object?> names = <Object?, Object?>{
    for (final Json p in people.values) p['name']: p['id'],
  };
  for (final Json p in people.values) {
    p['aliases'] =
        (p['aliases']! as List<String>)
            .where(
              (String x) =>
                  x.isNotEmpty &&
                  x != p['name'] &&
                  !generic.contains(x) &&
                  !generic.contains(x.replaceFirst(RegExp('^[他她的老小未]+'), '')) &&
                  !RegExp('的|夫妇|夫妻|们|一家|俩').hasMatch(x) &&
                  !(names.containsKey(x) && names[x] != p['id']),
            )
            .toSet()
            .toList();
    for (final MapEntry<String, Object?> e
        in (p['attr_history']! as Json).entries) {
      final List<Json> values = e.value! as List<Json>;
      // Preserve Python's stable order when two records share a position.
      final List<(int, Json)> sorted = <(int, Json)>[
        for (int i = 0; i < values.length; i++) (i, values[i]),
      ]..sort(((int, Json) a, (int, Json) b) {
        final int byPos = (a.$2['p']! as num).compareTo(b.$2['p']! as num);
        return byPos != 0 ? byPos : a.$1.compareTo(b.$1);
      });
      values
        ..clear()
        ..addAll(sorted.map(((int, Json) row) => row.$2));
      (p['attrs']! as Json)[e.key] = sorted.last.$2['v'];
    }
  }
  for (final Json e in events) {
    for (final String id
        in (e['who']! as List<Object?>).cast<String>().map(canon).toSet()) {
      if (people.containsKey(id)) (people[id]!['events']! as List<Json>).add(e);
    }
  }
  final Map<String, Json> rels = <String, Json>{};
  for (final Json r in relations) {
    final String a = canon(r['a']! as String), b = canon(r['b']! as String);
    if (a == b || !people.containsKey(a) || !people.containsKey(b)) continue;
    final List<String> pair = <String>[a, b]..sort();
    final String key =
        '${pair.join('|')}${r['family'] != null && r['family'] != '' ? '|${r['family']}' : ''}';
    final Json row = <String, Object?>{...r, 'a': a, 'b': b};
    if (a.compareTo(b) > 0)
      row.addAll(<String, Object?>{
        'a': b,
        'b': a,
        'a_is': r['b_is'] ?? '',
        'b_is': r['a_is'] ?? '',
      });
    final Json? previous = rels[key];
    if (previous != null) {
      final bool reverse = previous['a'] != row['a'];
      for (final String field in <String>['a_is', 'b_is', 'desc', 'status']) {
        final String old =
            reverse && field == 'a_is'
                ? 'b_is'
                : reverse && field == 'b_is'
                ? 'a_is'
                : field;
        if (row[field] == null || row[field] == '')
          row[field] = previous[old] ?? '';
      }
      row['history'] = <Json>[
        ...(previous['history']! as List<Json>),
        <String, Object?>{
          for (final MapEntry<String, Object?> e in previous.entries)
            if (e.key != 'history') e.key: e.value,
        },
      ];
    } else {
      row['history'] = <Json>[];
    }
    rels[key] = row;
  }
  return <String, Object?>{
    'people': people,
    'rels': rels,
    'events': events,
    'saga': saga,
    'recaps': recaps,
  };
}
